src/rdydata.c

```diff
@@ -23,60 +23,61 @@ static PetscErrorCode CheckBoundaryConditionID(RDy rdy, PetscInt bnd_cond_id) {
+  PetscCheck(ndof == 3, rdy->comm, PETSC_ERR_USER, "The number of DOFs (%d) for the boundary condition need to be three.", ndof);
 
-  RDyBoundary *boundary = &rdy->boundaries[bnd_cond_id];
-  PetscCheck(boundary->num_edges == num_edges, rdy->comm, PETSC_ERR_USER,
+  RDyBoundary boundary = rdy->boundaries[boundary_id];
+  PetscCheck(boundary.num_edges == num_edges, rdy->comm, PETSC_ERR_USER,
              "The number of edges (%d) in the data to set boundary condition for ID = %d, do not match the actual number of edge (%d)", num_edges,
-             bnd_cond_id, boundary->num_edges);
-
-  RDyCondition *boundary_cond = &rdy->boundary_conditions[bnd_cond_id];
-  PetscCheck(boundary_cond->flow->type == CONDITION_DIRICHLET, rdy->comm, PETSC_ERR_USER,
-             "Trying to set dirichlet values for boundary condition (%d), but it is of a different type (%d)", bnd_cond_id,
-             boundary_cond->flow->type);
-
-  RDyCells            *cells    = &rdy->mesh.cells;
-  RDyEdges            *edges    = &rdy->mesh.edges;
-  PetscRiemannDataSWE *data_swe = rdy->petsc_rhs;
-  RiemannDataSWE      *datar    = &data_swe->datar_bnd_edges[bnd_cond_id];
-  PetscReal            tiny_h   = rdy->config.physics.flow.tiny_h;
-
-  for (PetscInt e = 0; e < boundary->num_edges; ++e) {
-    PetscInt iedge = boundary->edge_ids[e];
-    PetscInt icell = edges->cell_ids[2 * iedge];
-
-    if (cells->is_local[icell]) {
-      datar->h[e] = bc_values[e * ndof];
-
-      if (datar->h[e] > tiny_h) {
-        datar->u[e] = bc_values[e * ndof + 1] / datar->h[e];
-        datar->v[e] = bc_values[e * ndof + 2] / datar->h[e];
-      } else {
-        datar->u[e] = 0.0;
-        datar->v[e] = 0.0;
-      }
+             boundary_id, boundary.num_edges);
+
+  RDyCondition boundary_cond = rdy->boundary_conditions[boundary_id];
+  PetscCheck(boundary_cond.flow->type == CONDITION_DIRICHLET, rdy->comm, PETSC_ERR_USER,
+             "Trying to set dirichlet values for boundary %d, but it has a different type (%d)", boundary_id, boundary_cond.flow->type);
+
+  // dispatch this call to CEED or PETSc
+  PetscReal tiny_h = rdy->config.physics.flow.tiny_h;
+  if (rdy->ceed_resource[0]) {  // ceed
+    // fetch the array storing the boundary values
+    CeedOperatorField dirichlet_field;
+    PetscCall(GetSWEFluxOperatorDirichletBoundaryValues(rdy->ceed_rhs.op_edges, boundary_id, &dirichlet_field));
+    CeedVector dirichlet_vector;
+    CeedOperatorFieldGetVector(dirichlet_field, &dirichlet_vector);
+    PetscInt num_comp = 3;
+    CeedScalar(*dirichlet_ceed)[num_comp];
+    CeedVectorGetArray(dirichlet_vector, CEED_MEM_HOST, (CeedScalar **)&dirichlet_ceed);
+
+    // set the boundary values
+    for (PetscInt i = 0; i < boundary.num_edges; ++i) {
+      dirichlet_ceed[i][0] = bc_values[num_comp * i];
+      dirichlet_ceed[i][1] = bc_values[num_comp * i + 1];
+      dirichlet_ceed[i][2] = bc_values[num_comp * i + 2];
```

Comment:
This is pretty neat and I think we should implement a similar code for [RDySetWaterSource()](https://github.com/RDycore/RDycore/blob/93285b4303002a000f127e81310197c28f9771b0/src/rdydata.c#L121-L131)

Reply:
Yeah, I was thinking the same thing. It makes our interface a bit smaller. I could do that in this PR, in fact, after I make sure we're actually testing the code that sets the Dirichlet BC values.